docs: refine documentation 0.7 (#643)
jina-ai/finetuner: docs/advanced-topics/linear-probe.md

(linear-probe)=
# {octicon}`pin` Linear Probe

## Why freezing?

Depending on your task and the amount of training data,

> sometimes it is unnecessary to tune the entire model.

Reviewer: ```suggestion
it is not always necessary to tune the entire model.
```

> In these cases,

Reviewer: ```suggestion
In some cases,
```

> freezing the weights of the pre-trained model and fine-tuning specific layers is sufficient enough.

Reviewer: ```suggestion
freezing some of the weights of the pre-trained model and just fine-tuning specific layers produces comparable or better results.
```

> Besides, freezing can reduce the training time considerably.

Reviewer: ```suggestion
Furthermore, freezing weights can reduce the training time dramatically.
```

> Finetuner allows you to fine-tune a Linear Probe (or Projection Head) easily.

Reviewer: Is there a wikipedia page or paper for this? I don't know what that is.

Author: different ppl use different words, for example: https://arxiv.org/pdf/2002.05709.pdf

Author: updated everything to projection head

```{warning}

> Currently, we only allow you to freeze layers for image-to-image search tasks.

Reviewer: I'm a little confused why this is true. Is it just that some kinds of models are supported?

Author: yes we only enabled it for CNNs, not transformers. It is supported inside Core, not Client

Reviewer 2: TODO:  make it clear why only image-to-image

> These models are built on top of Convolution Neural Networks (CNNs).

Reviewer: ```suggestion
These models are built on top of Convolutional Neural Networks (CNNs).
```

> For transformer architecture,

Reviewer: ```suggestion
For transformer architectures,
```

> we always train the entire Neural Nets.

Reviewer: ```suggestion
we can only fine-tune the entire neural network.
```

> If this feature is needed, consider submitting a feature request in our [Github Issues page](https://github.com/jina-ai/finetuner/issues)

Reviewer: ```suggestion
If you need to freeze weights for transformers, consider submitting a feature request in our [Github Issues page](https://github.com/jina-ai/finetuner/issues)
```

```

## How?

Finetuner has a built-in module called Tailor.
Given a general model written in Pytorch,

> Tailor performs micro-operations on the model architecture and outputs an embedding model.

Reviewer: ```suggestion
Tailor performs the micro-operations on the model architecture required for fine-tuning and outputs an embedding model.
```

> Given a general model with weights, Tailor preserves its weights and performs (some of) the following steps:

Reviewer: ```suggestion
Given a general model with weights, Tailor performs some or all of the following steps:
```

> + Finding all dense layers by iterating over layers.

Reviewer: ```suggestion
+ Iterating over all layers to find dense layers.
```

> + Chopping off all layers after a certain dense layer.

Reviewer: ```suggestion
+ Removing the layers after a certain dense layer.
```

Author: client supports freezing and attaching, and part of "chopping" because all FCs from CNNs has been chopped off

> + Freezing weights of specific layers.

Reviewer: ```suggestion
+ Freezing weights on specific layers.
```

> + Adding a new module on top of the embedding model.

Reviewer: ```suggestion
+ Adding a new layers on top of the model.
```

Author: a pytorch module = layers

![tailor](../imgs/tailor.svg)

> With Finetuner, you can simply put `freeze=True` and `output_dim=X` like this:

Reviewer: For example, just using the arguments `freeze=True` and `output_dim=X` with the `fit` function, as shown below:

> With Finetuner, you can simply put `freeze=True` and `output_dim=X` like this:

Reviewer: ```suggestion
For example, just using the arguments `freeze=True` and `output_dim=X` with the `fit` function, as shown below:
```

```diff
run = finetuner.fit(
    model='resnet50',
    ...,
+   freeze=True,
+   output_dim=1024,  # default output_dim of ResNet50 is 2048.
    ...,
)
```

Finetuner will:

1. Remove the classification head of `ResNet` model, and convert it into an embedding model.
2. Freeze all layers of the embedding model.
3. Attach a trainable 3-layer Linear Probe on top of the embedding model with an `output_dim=1024`.

## Summary

> If you want to achieve efficient fine-tuning without training the entire model,

Reviewer: ```suggestion
If you want to achieve efficient fine-tuning without retraining the entire model,
```

> tuning a Linear Probe could be a good solution.

Reviewer: ```suggestion
turning the model into a Linear Probe could be a good solution.
```

> Keep in mind that whenever you use `freeze=True`, always set the `output_dim`.

Reviewer: ```suggestion
Keep in mind that whenever you use `freeze=True`, always set `output_dim`.
```

> Otherwise, nothing can be tuned since all layers are frozen.

Author: i realized it is wrong. If `freeze=True`, we also attach a head. I'll update this line

> Otherwise, nothing can be tuned since all layers are frozen.

Author: fixed
